// @plasma-platform/quark-library/.babelrc.js
const path = require('path');
const projectPath = path.resolve(__dirname, './');

module.exports = {
  "presets": [
    ["@babel/preset-env", {
        "targets": {
          "node": "current",
          "browsers": ["> 3%", "ie 11"]
        },
        modules: 'umd',
        debug: false,
        forceAllTransforms: true,
        useBuiltIns: 'entry',
        loose: true,
    }],
    "@babel/preset-react",
    ["@babel/preset-stage-0", {
      "decoratorsLegacy": true,
    }]
  ],
  "plugins": [
    ["module-resolver", {
      "root": ["/"],
      "alias": {
        Utils: path.resolve(projectPath, 'src/utils/'),
        Components: path.resolve(projectPath, 'src/components/'),
        Images: path.resolve(projectPath, 'src/assets/images/'),
        Assets: path.resolve(projectPath, 'src/assets/'),
      }
    }],
    "@babel/plugin-proposal-export-default-from",
    "@babel/plugin-syntax-dynamic-import",
    "@babel/plugin-transform-runtime",
    ["@babel/plugin-proposal-decorators", {
      "legacy": true
    }],
    ["@babel/plugin-proposal-class-properties", {
      "loose": true
    }],
    "react-css-modules",
  ],
  "env": {
    "production": {
      "plugins": ["transform-react-remove-prop-types"]
    },
    "test": {},
  },
};
